docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,125 @@
+# Multi-tenancy with Vault Tokens
+
+## Current Feature: Vault access with single token
+
+The current feature to [support Hashicorp Vault as a KMS is
+documented](./encrypted-pvc.md):
+
+- tenants can create PVCs that are encrypted with a unique key
+- the key to encrypt/decrypt the PVC is stored in Hashicorp Vault
+- there is a single token (Kubernetes ServiceAccount) to access Vault
+- enabling is done in the StorageClass, pointing to a KMS configuration section
+  in a JSON configuration file (Kubernetes ConfigMap)
+
+## High-Level Requirements
+
+The new feature should support multi-tenancy, so that each Tenant can use their
+own Vault Token for PVC encryption:
+
+- each tenant can manage their Vault token (and PVC keys)
+- to get the key from Vault, each tenant should provide their personal token
+- Ceph-CSI should use the token from the tenant to store the unique key for the
+  PVC
+- a Tenant is the owner of a Kubernetes Namespace
+
+## Restrictions to consider
+
+- tenants can only configure their Kubernetes namespace, their token to access
+  Vault should be located in their namespace (as a Kubernetes Secret)
+- Ceph-CSI can not directly access Kubernetes namespaces, as CSI is an
+  abstraction layer
+- Ceph-CSI needs to talk to a service (sidecar) to access the token from a
+  tenant
+- the KMS configuration is a ConfigMap with name
+  `ceph-csi-encryption-kms-config` in the namespace where the Ceph-CSI pods are
+  running
+- the KMS configuration is available for the Ceph-CSI pods at
+  `/etc/ceph-csi-encryption-kms-config/config.json`
+- [example of the
+  configuration](https://github.com/ceph/ceph-csi/blob/master/examples/kms/vault/kms-config.yaml)
+
+## Dependencies
+
+- the name of the Kubernetes Secret needs to be known (configured in the
+  StorageClass, or somewhere else)
+- each tenant (Kubernetes namespace) should use the same name for the Secret
+  that contains the Vault token
+- the Ceph-CSI sidecar needs to store the namespace (and name of the Secret?)
+  together with the metadata of the PVC
+
+## Implementation Outline
+
+- when creating the PVC the Ceph-CSI provisioner needs to store the namespace
+  (and Kubernetes Secret?) of the PVC in its metadata
+  - stores the `csi.volume.owner` (name of Tenant) in the metadata of the
+    volume and sets it as `rbdVolume.Owner`
+- the Ceph-CSI node-plugin needs to request the token in the NodeStage CSI
+  operation and create/get the key for the PVC
+- the Ceph-CSI provisioner needs to request the token to delete the key for the
```

Comment:
The typical delete time issue of originating namespace and hence the token in a secret may be unavailable at delete time, i.e the namespace itself may have been deleted and the delete volume would come in asynchronously.

This is unavoidable, so the delete operation would not be able to garbage collect the key for this PVC? If so, when the same PVC is recreated do we intend to reuse the existing key or a new one (if it is possible to detect that the existing key is stale)?

Reply:
That is a good point. There is a dependency on the Secret in the Namespace of the Tenant. When the Namespace is deleted, there is no defined order in the deletion of the Secret or PVCs. Deleting the passphrase for the PV will not be possible when the Secret has been removed. Not sure how this can be avoided or solved.

For the recreation, maybe it is not sufficient to have the owner, but it also needs a UUID of the Kubernetes Namespace (or other unique ID of the tenant). However, people may want to restore their Namespace + PVCs from backup, and in that case we would want to use the pre-existing passphrases again...

Reply:
Again, let's park this in the "Gaps" section.

@nixpanic do we want to add that section and then merge this in?

Reply:
Updated the doc with a reference to #1782.